### PR TITLE
DRILL-8434: Add Median Function

### PR DESCRIPTION
**Pull request by cgivre**
# [DRILL-8434](https://issues.apache.org/jira/browse/DRILL-8434): Add Median Function

## Description
Adds a UDF to calculate the median of a column.

This PR also removes the `Deprecated` tag from the `ObjectHolder` since that is in wide usage throughout Drill and we have no plans or work to refactor that any time soon.

Usage:
`median(n)`

## Documentation
See above...

## Testing
Added unit tests

### CODE REVIEW

File: exec/vector/src/main/java/org/apache/drill/exec/expr/holders/ObjectHolder.java [hunk 1]
```diff
@@ -26,7 +26,6 @@
  * This holder can be used only as workspace variables in aggregate functions.
  * Using this holder should be avoided and we should stick to native holder types.
  */
-@Deprecated
```
Comment (jnturton) [inline exec/vector/src/main/java/org/apache/drill/exec/expr/holders/ObjectHolder.java:29]: Can we delay this removal a bit?

File: contrib/udfs/src/main/java/org/apache/drill/exec/udfs/DistributionFunctions.java [hunk 2]
```diff
@@ -363,4 +365,44 @@ public void eval() {
       }
     }
   }
+  @SuppressWarnings("unused")
+  @FunctionTemplate(name = "median",
```
Comment (jnturton) [inline contrib/udfs/src/main/java/org/apache/drill/exec/udfs/DistributionFunctions.java:369]: Have you considered using the same Freemarker template approach followed by the other aggregate statistical functions? If you can rebase on #2804 then I'm hopeful that any preferred cast ambiguity for a VARCHAR argument will be resolved.

Comment (cgivre) [reply contrib/udfs/src/main/java/org/apache/drill/exec/udfs/DistributionFunctions.java]: @jnturton I went ahead and did this with Freemarker templates as I think the approach is generally better anyway.

File: exec/vector/src/main/java/org/apache/drill/exec/expr/holders/ObjectHolder.java (hunk 1 above)
Comment (cgivre) [reply exec/vector/src/main/java/org/apache/drill/exec/expr/holders/ObjectHolder.java:29]: Sure I suppose.

Comment (cgivre) [reply exec/vector/src/main/java/org/apache/drill/exec/expr/holders/ObjectHolder.java:29]: Fixed

**Comment by cgivre** [comment]
@jnturton I reverted the deprecation for this.  Any other comments?

**Review by jnturton — APPROVED** [review]
Thanks +1.

File: exec/java-exec/src/main/codegen/templates/MedianFunctions.java.bak [hunk 3]
```diff
@@ -0,0 +1,88 @@
+/*
```
[the remark this answers is not in the document]
Comment (cgivre) [reply exec/java-exec/src/main/codegen/templates/MedianFunctions.java.bak]: Oops... removed.

**Pull request merged**